src/java.base/share/classes/java/lang/Runtime.java

```diff
@@ -45,10 +44,66 @@
+ *
+ * <h2><a id="shutdown">Shutdown Sequence</a></h2>
+ *
+ * <p>The Java Virtual Machine initiates the <i>shutdown sequence</i> in response
+ * to one of several events:
+ * <ol>
+ * <li>when the number of {@linkplain Thread#isAlive() live} non-daemon threads drops to zero
+ * for the first time;</li>
+ * <li>when the {@link #exit Runtime.exit} or {@link System#exit System.exit} method is called
+ * for the first time; or</li>
+ * <li>when some external event occurs, such as an interrupt or a signal is received from
+ * the operating system.</li>
+ * </ol>
+ *
+ * <p>At the beginning of the shutdown sequence, the registered shutdown hooks are
+ * {@linkplain Thread#start started} in some unspecified order. They run concurrently
+ * with any daemon or non-daemon threads that were {@linkplain Thread#isAlive() alive}
+ * at the beginning of the shutdown sequence.
+ *
+ * <p>After the shutdown sequence has begun, registration and de-registration of shutdown hooks
+ * with {@link #addShutdownHook addShutdownHook} and {@link #removeShutdownHook removeShutdownHook}
+ * is prohibited. However, creating and starting new threads is permitted. New threads run
+ * concurrently with the registered shutdown hooks and with any daemon or non-daemon threads
+ * that are already running.
+ *
+ * <p>The shutdown sequence finishes when all shutdown hooks have terminated. At this point,
+ * the Java Virtual Machine terminates as described below.
+ *
+ * <p>It is possible that one or more shutdown hooks do not terminate, for example, because
+ * of an infinite loop. In this case, the shutdown sequence will never finish. Other threads
+ * and shutdown hooks continue to run and can terminate the JVM via the {@link #halt halt} method.
+ *
+ * <p>Prior to the beginning of the shutdown sequence, it is possible for a program to start
+ * a shutdown hook by calling its {@link Thread#start start} method explicitly. If this occurs, the
+ * effect on the shutdown sequence is unspecified.
+ *
+ * <h2><a id="termination">Virtual Machine Termination</a></h2>
+ *
+ * <p>The JVM terminates when the shutdown sequence finishes or when {@link #halt halt} is called.
+ *
+ * <p>When the JVM terminates, all threads are immediately prevented from executing any further
+ * Java code. This includes shutdown hooks as well as daemon and non-daemon threads. The
+ * threads' current methods do not complete normally or abruptly; no {@code finally} clause
+ * of any method is executed, nor is any {@linkplain Thread.UncaughtExceptionHandler
+ * uncaught exception handler}.
+ *
+ * @implNote
+ * Conventionally, the native code that has launched the JVM will invoke the JNI function
+ * <a href="{@docRoot}/../specs/jni/invocation.html#destroyjavavm">DestroyJavaVM</a>.
+ * This function is responsible for initiating the shutdown sequence when the number of running
+ * ({@linkplain Thread#isAlive() live}) non-daemon threads first drops to zero. When the shutdown
```

Comment:
I think we need to align this more with the proposed text of JLS 12.8. It needs to be clearer that `DestroyJavaVM` must be used in conjunction with `CreateJavaVM` to get the desired termination behaviour. Otherwise, the proposed text allows for the possibility that the JVM will always terminate when the last non-daemon thread terminates, even if `DestroyJavaVM` has not been called. It needs to be clear that if the JVM is started by the JNI Invocation API, then it must also be terminated by the JNI invocationAPI to get the required behaviour.

Reply:
OK, yes I accept the general premise of your comment. Indeed we need to make sure this section is well aligned with JLS 12.8 (as well as the JNI spec and the JVMS, though I think those can be a bit more distant). I see that Alex has updated JLS 12.8 more recently than your comment, so I'll hold off editing this section for now and check with him to make sure he's done editing before I proceed. We might need some additional discussion as well. So, more to come.

Reply:
@dholmes-ora OK I've updated the implementation note to try to make it quite clear that the shutdown-at-zero-non-daemon-threads behavior is tied to `DestroyJavaVM`. (This is good; I didn't know this before I embarked on this exercise of updating these specs!) I'm not entirely sure of the state of JLS 12.8, but if I know Alex :-) it's unlikely to change much, and I think this section is consistent with it.